google-cloud-storage/lib/google/cloud/storage/file.rb

```diff
@@ -1199,11 +1199,11 @@ def rewrite dest_bucket_or_path, dest_path = nil,
 
           update_gapi = nil
           if block_given?
-            updater = Updater.new gapi
+            updater = Updater.new gapi.dup
```

Comment:
@frankyn Do you think there is any danger in passing the *complete* representation of the `sourceObject` as the request body (Object resource) to `objects.rewrite`? This resource would presumably contain `bucket`, `name`, etc as well as the desired "metadata" fields such as `cacheControl`, `contentType`, etc.

Reply:
I've tested this solution and it seems to work as desired, but the actual internal behavior of `objects.rewrite` is opaque.

Reply:
How much metadata was being copied in the past? 

I'd rather not magically copy metadata if the user didn't request the copy. @tseaver I think this is similar to the ACL copy in client level.

Reply:
@frankyn PTAL at the code example in #4254 for current behavior. Our users via shrine-google_cloud_storage find the current behavior of not copying fields such as `content-type` to be problematic.

Reply:
> How much metadata was being copied in the past? I'd rather not magically copy metadata if the user didn't request the copy.

@renchap Any comment?

Reply:
The thing here is that `File#copy` has an inconsistent behaviour. When you pass it a block then metadata are not copied, but if you do not pass a block then they are copied.
I think most of the time people want the metadata to be copied as well, but maybe we can make the default behavior consistent (either always copy, or never copy), and use a method argument to enable/disable metadata copy?

Reply:
@frankyn: My observation is that @renchap is correct in that if **no block** is provided, the client does **not** pass an object resource as the request body, and the GCS service copies the object with all metadata intact. Only when the the client passes a partial, "PATCH-style" object resource with only *changed* fields populated (in the current block behavior) does the GCS service copy the object **without** all metadata intact. Since the default behavior (no object resource) of [Objects: rewrite](https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite) is to copy all metadata, this behavior can be presumed to be the correct one. My conclusion is that Objects: rewrite was never intended to accept partial, "PATCH-style" object resource representations. Therefore I feel the change in this PR is correct.

Reply:
I added a [commit](https://github.com/googleapis/google-cloud-ruby/pull/4275/commits/440b78ee010944c54bf9a88d73f77b5a48c60869) demonstrating that if no block is provided, the GCS service copies the object with all metadata intact.

Reply:
Hi folks, I appreciate your patience.

@quartzmo and @renchap, thanks for clarifying.

Objects: rewrite, will only copy object metadata when a resource body isn't provided. If a resource body is provided GCS considers this the expected result which acts as an UPDATE rather than a PATCH, e.g. only modify the parts but keep the rest.

I agree that it's confusing that the rest of the metadata isn't copied, but if if the client forces a PATCH style version of rewrite, behavior expectations will change for the methods File#copy and File#rewrite. Existing users might dependent on this behavior and updating it could break them. 

I'd recommend a flag to copy existing metadata and allow it be modified with a block. In this way the actual API behavior is transparent to users and a user can clearly opt-into this type of request.

```ruby
require "google/cloud/storage"

gcs = Google::Cloud::Storage.new(project: '##project')
bucket_name = "##bucket_name"

bucket = gcs.bucket(b, skip_lookup: true)

original_file = bucket.file("test.png")

original_file.content_type

# force_copy_metadata - copy existing metadata which can be used in tandem with a block to perform a PATCH like request.
#           If no changes to the source object metadata is required then Cloud Storage will copy metadata.

copied_file = original_file.copy(bucket_name, "test_2.png", force_copy_metadata: true) do |f|
   f.cache_control = "public, max-age: 7200"
end
``` 


What do you both think?

Reply:
An opt-in flag sounds good to me. @renchap?

Reply:
Looks good as well, as well as a clear explanation of the behaviour in the documentation of the gem, so people are not confused by what is copied and when (even if this matches the standard GCS behaviour, all users of the Ruby gem may not know the small details of GCS)

Reply:
Great, I'll add a commit adding documentation of the default behavior, and also adding the flag `force_copy_metadata` to `File#copy` (but not `File#rewrite` unless someone has a reason to do so).

Reply:
@renchap @frankyn I added `force_copy_metadata` and updated docs in `File#copy` and `#rewrite`.

Reply:
@frankyn PTAL